guppylang/decorator.py

```diff
@@ -104,89 +115,137 @@ def _get_python_caller(self, fn: PyFunc | None = None) -> ModuleIdentifier:
 
         def dec(c: type) -> type:
-            module._register_buffered_instance_funcs(defn)
+            mod._register_buffered_instance_funcs(defn)
             return c
 
         return dec
 
     @pretty_errors
     def type(
         self,
-        module: GuppyModule,
         hugr_ty: ht.Type,
         name: str = "",
         linear: bool = False,
         bound: ht.TypeBound | None = None,
+        module: GuppyModule | None = None,
     ) -> OpaqueTypeDecorator:
         """Decorator to annotate a class definitions as Guppy types.
 
         Requires the static Hugr translation of the type. Additionally, the type can be
         marked as linear. All `@guppy` annotated functions on the class are turned into
         instance functions.
         """
-        module._instance_func_buffer = {}
+        mod = module or self.get_module()
+        mod._instance_func_buffer = {}
 
         def dec(c: type) -> OpaqueTypeDef:
             defn = OpaqueTypeDef(
-                DefId.fresh(module),
+                DefId.fresh(mod),
                 name or c.__name__,
                 None,
                 [],
                 linear,
                 lambda _: hugr_ty,
                 bound,
             )
-            module.register_def(defn)
-            module._register_buffered_instance_funcs(defn)
+            mod.register_def(defn)
+            mod._register_buffered_instance_funcs(defn)
             return defn
 
         return dec
 
-    @pretty_errors
-    def struct(self, module: GuppyModule) -> StructDecorator:
+    @property
+    def struct(
+        self,
+    ) -> Callable[[PyClass | GuppyModule], StructDecorator | RawStructDef]:
         """Decorator to define a new struct."""
-        module._instance_func_buffer = {}
+        # Note that this is a property. Thus, the code below is executed *before*
+        # the members of the decorated class are executed.
+        # At this point, we don't know if the user has called `@struct(module)` or
+        # just `@struct`. To be safe, we initialise the method buffer of the implicit
+        # module either way
+        caller_id = self._get_python_caller()
+        implicit_module_existed = caller_id in self._modules
+        implicit_module = self.get_module(
+            # But don't try to do implicit imports since we're not sure if this is
+            # actually an implicit module
+            resolve_implicit_imports=False
+        )
+        implicit_module._instance_func_buffer = {}
 
-        def dec(cls: type) -> RawStructDef:
+        def dec(cls: type, module: GuppyModule) -> RawStructDef:
             defn = RawStructDef(DefId.fresh(module), cls.__name__, None, cls)
             module.register_def(defn)
             module._register_buffered_instance_funcs(defn)
+            # If we mistakenly initialised the method buffer of the implicit module
+            # we can just clear it here
+            if module != implicit_module:
```

Comment:
this is doing pointer equality on the module then?

Reply:
Yes, since `GuppyModule` doesn't implement `__eq__`